sdk/core/System.BinaryData/src/BinaryData.cs

```diff
@@ -133,13 +129,22 @@ public BinaryData(string data)
             {
                 throw new ArgumentNullException(nameof(stream));
             }
-            if (stream.CanSeek && stream.Length > int.MaxValue)
+            if (stream.CanSeek && (stream.Length - stream.Position) > int.MaxValue)
             {
                 throw new ArgumentOutOfRangeException(
                     nameof(stream),
                     "Stream length must be less than Int32.MaxValue");
             }
-            using var memoryStream = new MemoryStream();
+
+            MemoryStream? inputMemoryStream = stream as MemoryStream;
+            ArraySegment<byte> buffer = default;
+
+            if (inputMemoryStream?.TryGetBuffer(out buffer) ?? false)
+            {
+                return new BinaryData(buffer.Array.AsMemory(buffer.Offset, buffer.Count));
+            }
+
+            using MemoryStream memoryStream = stream.CanSeek ? new MemoryStream((int)stream.Length) : new MemoryStream();
```

Comment:
`(int)stream.Length` can overflow, but `(int)(stream.Length - stream.Position)` can't (since it was already checked).

My recommendation is to create `int streamLength` local, break apart the and from the stream.CanSeek test to nested ifs, and assign streamLength inside that if, rather than repeating the subtraction.

```C#
int streamLength;

if (stream.CanSeek)
{
    long longLength = stream.Length - stream.Position;

    if (longLength > int.MaxValue)
    {
        throw new ...
    }

    streamLength = (int)longLength;
}
```